farm/data_handler/processor.py

```diff
@@ -1727,6 +1728,292 @@ def _sample_to_features(self, sample) -> dict:
+                                            'label': 'hard_negative',
+                                            'external_id': str},
+                                            {'title': str,
+                                            'text': str,
+                                            'label': 'positive',
+                                            'external_id': str},
+                                            ....
+                                            ]
+                            }
+
+        Returns:
+                sample: instance of Sample
+        """
+        # extract query, positive context passages and titles, hard-negative passages and titles
+        query = self._normalize_question(dictionary["query"])
+        positive_context = list(filter(lambda x: x["label"] == "positive", dictionary["passages"]))
+        if self.shuffle_positives:
+            random.shuffle(positive_context)
+        positive_context = positive_context[:self.num_positives]
+        hard_negative_context = list(filter(lambda x: x["label"] == "hard_negative", dictionary["passages"]))
+        if self.shuffle_negatives:
+            random.shuffle(hard_negative_context)
+        hard_negative_context = hard_negative_context[:self.num_hard_negatives]
+
+        positive_ctx_titles = [passage.get("title", None) for passage in positive_context]
+        positive_ctx_texts = [passage["text"] for passage in positive_context]
+        hard_negative_ctx_titles = [passage.get("title", None) for passage in hard_negative_context]
+        hard_negative_ctx_texts = [passage["text"] for passage in hard_negative_context]
+
+        # all context passages and labels: 1 for positive context and 0 for hard-negative context
+        ctx_label = [1]*self.num_positives + [0]*self.num_hard_negatives #(self.num_positives if self.num_positives < len(positive_context) else len(positive_context)) + \
+        # +(self.num_hard_negatives if self.num_hard_negatives < len(hard_negative_context) else len(hard_negative_context))
+
+        # featurize the query
+        query_inputs = self.query_tokenizer.encode_plus(
+            text=query,
+            max_length=self.max_seq_len,
+            add_special_tokens=True,
+            truncation_strategy='do_not_truncate',
+            padding="max_length",
+            return_token_type_ids=True,
+        )
+
+        # featurize context passages
+        if self.embed_title:
+            # embed title with positive context passages + negative context passages
+            all_ctx = [tuple((title, ctx)) for title, ctx in
+                       zip(positive_ctx_titles, positive_ctx_texts)] + \
+                      [tuple((title, ctx)) for title, ctx in
+                       zip(hard_negative_ctx_titles, hard_negative_ctx_texts)]
+        else:
+            all_ctx = positive_ctx_texts + hard_negative_ctx_texts
+
+        # assign empty string tuples if hard_negative passages less than num_hard_negatives
```

Comment:
Add a small comment why we need this. I believe it's because we want to have the same dimensionalities of tensors in our batches later on(?)